core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -338,105 +214,56 @@ private void setResponse(BulkItemRequest request, BulkItemResponse response) {
+                                                                         int requestIndex, UpdateRequest updateRequest)
+        throws Exception {
+        // Todo: capture read version conflicts, missing documents and malformed script errors in the write result due to get request
         UpdateHelper.Result translate = updateHelper.prepare(updateRequest, indexShard);
         switch (translate.getResponseResult()) {
             case CREATED:
             case UPDATED:
                 IndexRequest indexRequest = translate.action();
-                try {
-                    WriteResult result = shardIndexOperation(bulkShardRequest, indexRequest, metaData, indexShard, false);
-                    return new UpdateResult(translate, indexRequest, result);
-                } catch (Exception e) {
-                    final Throwable cause = ExceptionsHelper.unwrapCause(e);
-                    boolean retry = false;
-                    if (cause instanceof VersionConflictEngineException) {
-                        retry = true;
+                MappingMetaData mappingMd = metaData.mappingOrDefault(indexRequest.type());
+                indexRequest.process(mappingMd, allowIdGeneration, request.index());
+                WriteResult<IndexResponse> writeResult = TransportIndexAction.executeIndexRequestOnPrimary(indexRequest, indexShard, mappingUpdatedAction);
+                BytesReference indexSourceAsBytes = indexRequest.source();
+                IndexResponse indexResponse = writeResult.getResponse();
+                UpdateResponse update = new UpdateResponse(indexResponse.getShardInfo(), indexResponse.getShardId(), indexResponse.getType(), indexResponse.getId(), indexResponse.getVersion(), indexResponse.getResult());
+                if (translate.getResponseResult() == DocWriteResponse.Result.CREATED) {
+                    if ((updateRequest.fetchSource() != null && updateRequest.fetchSource().fetchSource()) ||
+                            (updateRequest.fields() != null && updateRequest.fields().length > 0)) {
+                        Tuple<XContentType, Map<String, Object>> sourceAndContent = XContentHelper.convertToMap(indexSourceAsBytes, true);
+                        update.setGetResult(updateHelper.extractGetResult(updateRequest, updateRequest.concreteIndex(), indexResponse.getVersion(), sourceAndContent.v2(), sourceAndContent.v1(), indexSourceAsBytes));
+                    } else {
+                        update.setGetResult(null);
                     }
-                    return new UpdateResult(translate, indexRequest, retry, cause, null);
+                } else {
+                    assert translate.getResponseResult() == DocWriteResponse.Result.UPDATED;
+                    update.setGetResult(updateHelper.extractGetResult(updateRequest, updateRequest.concreteIndex(), indexResponse.getVersion(), translate.updatedSourceAsMap(), translate.updateSourceContentType(), indexSourceAsBytes));
```

Comment:
why do we now return always the get result on updates? it seems to be different [before](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java#L251)